Tamper protected Endpoint uninstall - combined PR

aleksmaus opened the pull request:
## What does this PR do?

Combined 3PRs that cover all the use cases that lead to Endpoint uninstall from the following PRs:
[Pass uninstall token from Agent uninstall command to Endpoint uninstall command](https://github.com/elastic/elastic-agent/pull/2708)
[Support singed UNENROLL action and UNENROLL forwarding to Endpoint](https://github.com/elastic/elastic-agent/pull/2723)
[Tamper protected Endpoint integration removal](https://github.com/elastic/elastic-agent/pull/2747)

Opening as draft for now, not sure how and when this will get merged, but it's more convenient to test for Endpoint.
@intxgo 

## Why is it important?

Implements tamper protection for endpoint uninstall use cases.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.667% (`74/75`) | :+1: 
Files | 69.38% (`179/258`) | :+1: 1.163
Classes | 68.107% (`331/486`) | :+1: 0.891
Methods | 53.957% (`1009/1870`) | :+1: 0.384
Lines | 39.771% (`11579/29114`) | :+1: 0.274
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

aleksmaus commented:
Opening this PR for review. It consolidates 3 previously reviewed PRs in one PR.

The following manual tests where completed with:
1. Endpoint build:
https://endgame-ci.elastic.co/user/intxgo/my-views/view/Branches/job/elastic+endpoint-dev+multibranch+pipeline/job/lesio%252Funinstall_token/14/
2. Kibana branch:
https://github.com/joeypoon/kibana/pull/new/feature/sign-unenroll

The Kibana feature flag should be enabled as implemented in 
https://github.com/elastic/kibana/pull/158494
This shows the protection toggle on UI.
<img width="775" alt="Screenshot 2023-06-12 at 2 39 26 PM" src="https://github.com/elastic/elastic-agent/assets/872351/09eedc3d-a9c0-4a6c-a347-0793fe77eebc">

The uninstall tokens are not exposed on UI yet, but can be grabbed from the log, on that branch.
Kibana writes something like:
```
~~~~~~~~~~~~ tokens:
{
	"fc6c29f0-f8cb-11ed-ab07-d3e4064cb6ee":"e87c61db9808ff4df51692cb7f571885",
	"6314e110-f8c2-11ed-ab07-d3e4064cb6ee":"739ff6fb3f233f5d1519ec53b74d0578",
	"fleet-server-policy":"43810fede82621e9fb9baf32ce1b2fab"
}
```


[Pass uninstall token from Agent uninstall command to Endpoint uninstall command](https://github.com/elastic/elastic-agent/pull/2708)

1. With Agent protection disabled, the Agent and Endpoint uninstalls successfully without token.
2. With Agent protection disabled, the Agent and Endpoint uninstalls with any token value. The token validation is skipped.
3. With Agent protection enabled Endpoint fails to uninstall without token or invalid token.
4. With Agent protection enabled Endpoint uninstalls successfully with the valid token.

[Support singed UNENROLL action and UNENROLL forwarding to Endpoint](https://github.com/elastic/elastic-agent/pull/2723)

The UNENROLL doesn't uninstall the agent service this is known current behavior of the Agent.

1. With Agent and Endpoint protection disabled, the Agent can be UNENROLLED (regression test)
2. With Agent and Endpoint protection enabled, the Agent can be UNENROLLED.

Known behavior: It takes few moments for Endpoint to establish connection with the Agent and protect itself, thus Endpoint could be right uninstalled after it is started, before it gets protected. 

[Tamper protected Endpoint integration removal](https://github.com/elastic/elastic-agent/pull/2747) 
1. With Agent and Endpoint protection disabled, adding and removing Endpoint integration to the policy still works as expected (regression test)
2. With Agent and Endpoint protection enabled, the Endpoint is stopped and uninstalled successfully upon removal of Endpoint integration from the policy. 

The easiest way to check that the "tamper protection" was enabled for Endpoint is to attempt to uninstall it without uninstall token, you should get the error:
```
$ sudo /opt/Elastic/Endpoint/elastic-endpoint uninstall --log stderr
2023-06-12 17:48:32: debug: Utilities.cpp:308 Protected data {"agent":{"protection":{"enabled":true,"signing_key":"MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAENxwK08uYPM8D5yTmoA/emoxK5s5lQFyt/6E4LpMmpxbukMBN3/rNfjodisQBPyVQmE/ENuTbE42G5K8iXXvBRA==","uninstall_token_hash":"tx/Vht5/TVZ7gE8LcuX0NiV9azGfpX3BkrF0/wRjFsY="}},"id":"6314e110-f8c2-11ed-ab07-d3e4064cb6ee","inputs":[{"id":"7429fcd3-075b-474e-8f03-d3dcf98fdeb8","name":"EP","revision":1,"type":"endpoint"}]}
2023-06-12 17:48:32: error: MainPosix.cpp:188 Invalid uninstall token
```

AndersonQ reviewed (dismissed):
I haven't looked at the whole PR, but I left a comment, the one marked "[Blocker]", I believe needs to be addressed.

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

aleksmaus commented on internal/pkg/agent/application/actions/handlers/handler_action_unenroll.go:
```diff
@@ -40,18 +54,61 @@ type Unenroll struct {
+		}(i))
+	}
+
+	return g.Wait()
```
In my opinion it's appropriate to run the actions dispatch in parallel like this, this would return the first error and cancel context on the first error returned by any of the "Go routined" functions. This pattern allows early cancellation on error.
The current agent implementation has fairly few places that use the same ```errgroup``` approach. 
Also if you insist on returning the array of errors then the more context would need to be attached to every error in order to be able to correlate these errors to specific dispatch call, because the order of the errors will be random in general.

Please let me know if you feel strongly about returning an array of errors or the wrapper error containing that array instead and still want me to change the code.

aleksmaus commented on internal/pkg/agent/application/actions/handlers/handler_action_unenroll.go:
```diff
@@ -40,18 +54,61 @@ type Unenroll struct {
+						log.Warnf("UNENROLL failed for %v, err: %v", comp.ID, strErr)
+						return errors.New(strErr)
+					}
+				}
```
yep, updated.

aleksmaus commented:
Added support tamper protected upgrade.
Testing upgrade is quite tricky. Here is what I had to do:

1. Spin up local kibana (8.9.0) from Joey’s branch https://github.com/joeypoon/kibana/tree/feature/sign-unenroll
2. Build a 8.8.0 version of the agent with all the tamper protection, so I can test upgrade from tamper protected agent to 
tamper protected agent
3. Build the same code as 8.9.0 agent
4. Build and run a small local server that can serve the agent distros over http download curl http://localhost:8080/downloads/beats/elastic-agent/elastic-agent-8.9.0-linux-arm64.tar.gz.sha512
5. Change the upgrade download url in kibana to point to the local file server the hosts the 8.9.0 agent package
6. Install 8.8.0 agent against that local 8.9.0 stack
7. Ensure the agent is healthy on kibana UI
8. Kick off upgrade from 8.8.0 to 8.9.0 in this case, but now we are looking at more like upgrading from 8.11 to 8.12 realistically speaking
9. Ensure the upgrade succeeds in both case when the Endpoint is unprotected and when the Endpoint is protected.

aleksmaus commented:
Will be adding the feature flag here next. This should finalize covering all the requirements to this feature so far.

aleksmaus commented:
Added feature flag, retested everything with the latest Endpoint, including the feature flag transition.

It's largely ready for review and feedback.

In order to set the feature flag have to use API, http PUT to 
```
/api/fleet/agent_policies/<policy id>
```
with body payload, something like:
```
{
    "name": "Agent policy 1",
    "namespace": "default",
    "agent_features": [
        {
            "name": "tamper_protection",
            "enabled": true
        }
    ]
}
```
<img width="578" alt="Screenshot 2023-06-26 at 7 08 19 PM" src="https://github.com/elastic/elastic-agent/assets/872351/54d904ca-897d-4fdd-b955-2b9713a50201">



One issue that will be tracking down tomorrow with the Endpoint team: The Endpoint gets unprotected upon UPGRADE action, but is left unprotected, until the new policy comes in.

faec commented on pkg/component/input_spec.go:
```diff
@@ -12,13 +12,14 @@ import (
+	Description    string      `config:"description" yaml:"description" validate:"required"`
+	Platforms      []string    `config:"platforms" yaml:"platforms" validate:"required,min=1"`
+	Outputs        []string    `config:"outputs,omitempty" yaml:"outputs,omitempty"`
+	ProxiedActions []string    `config:"proxied_actions,omitempty" yaml:"proxied_actions,omitempty"`
```
This adds a new field to the input spec format -- please document this field in `docs/component-specs.md`, it's not obvious what its intended effect is

aleksmaus replied:
Done

faec commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,104 @@
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Iterate through found components and forward the UNENROLL action
```
Misleading comment? This code doesn't look specific to UNENROLL

aleksmaus replied:
Yes, was a result of refactoring the reusable function. Fixed.

faec commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,104 @@
+	// Iterate through found components and forward the UNENROLL action
+	for i := 0; i < len(comps); i++ {
+		g.Go(func(idx int) func() error {
+			comp := comps[i]
```
I can see by tracing through to `findMatchingUnitsByActionType`, the source of these arrays, that this is probably safe, but it's not obvious from the function definition or comments. Most arrays of components and units have no reason to be the same size, and the requirement for these parameters isn't documented.

Suggestion: define a helper like:
```go
type unitWithComponent struct {
  unit component.Unit
  component component.Component
}
```

then have `findMatchingUnits...` return an array of these, and `dispatchAction...` take that array as input, so you can just iterate directly over the list without relying on extra metadata about the list indexes. Then it's immediately obvious that the bounds checking is correct and that these values are intentionally paired.

aleksmaus replied:
Done.

faec commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,104 @@
+
+	// Iterate through found components and forward the UNENROLL action
+	for i := 0; i < len(comps); i++ {
+		g.Go(func(idx int) func() error {
```
Related to the comment immediately below: can you break this call up a bit, so you aren't invoking an async call on the invocation of another function returning a function all in one line? With the previous suggestion, you could define a standalone helper before the loop, like `dispatch := func(uc unitWithComponent) error { ... }`, then the for loop could be:

```go
for _,uc := range unitComponentPairs {
  // Send the action to the target unit via g.Go to collect any resulting errors
  target := uc
  g.Go(func() error { return dispatch(target) })
}
```

(variable names subject to taste 😜)

faec commented on pkg/component/component.go:
```diff
@@ -116,6 +174,12 @@ func (c *Component) Type() string {
 }
 
+// Model components model
+type Model struct {
```
Please add comments to this struct describing how it is used and the effects of each field. How does `Model.Signed.Data` related to `Model.Components`? Where does the signature in `Signed` come from, and which components use it? At what stage(s) is the signature verified -- do consumers of the `Components` field have any obligation to check anything?

aleksmaus replied:
Added.

faec replied:
Thank you, this helps a lot. Some followup questions: What form of the policy is signed / verified? For example:
- Some policies use variable substitution. If a policy field has `${some_variable}` in it, will that substitution break the signature?
- Policies can include conditions that are required for certain inputs or config subtrees, and if those conditions evaluate to false it can remove units or pieces of their config. Even if the condition passes, its raw text representation within the policy is altered. Does this break the signature?

Related: this description looks like the signature covers the full component list, is that right or does it split up componentwise? There are various things that might block a specific input (aside from the conditions mentioned above there may be platform restrictions / preventions, `capabilities.yml` blocks on certain input types, etc). If an unrelated input is removed from the model, would that break the signature from Endpoint's perspective?

aleksmaus replied:
Only small portion of the data from the policy is signed, in this care only data to support tamper protection use cases, for example:
```
{
    "id": "de8cf520-11fa-11ee-87c3-87540e933c0c",
    "agent": {
        "features": {
            "tamper_protection": {
                "enabled": true
            }
        },
        "protection": {
            "enabled": false,
            "uninstall_token_hash": "hF8DWoc2hdb9jm0NAtNwQFVpb6FUCtkIquS/IDtO8jE=",
            "signing_key": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEptaH31UPWD4b9mo+W8uzKnjN+w3aIWMmoxuR3qQsXs8oRfKQ+8AhdDHHMIWCeNz++HAxu5J3srPAq3vobND6wQ=="
        }
    },
    "inputs": [
        {
            "id": "5be5a402-b3cb-456a-92e7-5ca5ac9204af",
            "name": "EP",
            "revision": 1,
            "type": "endpoint"
        }
    ]
}
```
The reason the whole policy can't be signed because some policy fields are already updated by the fleet server.

Replacement of the values in the policy will not not break the signature, since the data (```signed.data``` field) is serialized and signed before the policy saved in Elasticsearch by kibana and is never modified between kibana and Endpoint for example. 

Over the time we can carefully sign more data from the policy, but we would have to continue support the the legacy behavior for some time, thus will not be able to protect the whole policy from tampering.

faec reviewed (commented):
Gave an initial pass -- this is pretty big and I'd like to understand its behavior better before approving, so the main thing I'm looking for is more documentation / explanation to help evaluate it.

aleksmaus commented:
@faec Thanks for the feedback. I addressed your comments.
The links in the description point to the original PRs and to the original issue
https://github.com/elastic/security-team/issues/5798

Don't hesitate to reach out with any questions.
Thanks!

faec commented on docs/component-specs.md:
```diff
@@ -53,6 +53,17 @@ The platforms this input or shipper supports. Must contain one or more of the fo
 
 The output types this input or shipper supports. If this is an input, then inputs of this type can only target (non-shipper) output types in this list. If this is a shipper, then this shipper can only implement output types in this list.
 
+### `proxied_actions` (list of strings)
```
Sorry, could you elaborate here a bit more? It's not obvious to me what it would mean to "proxy an action to the corresponding component" -- my impression from looking at the code is that the component is being notified of the action, but I'm not sure what makes it a proxy.

If I was guessing based on the code, I'd call this something like `forwarded_actions` or `notify_actions`. When I see "proxy" I think of something acting as direct intermediary for an action, or masking the identity of the sender/receiver. Is my intuition here right or is "proxy" meant to suggest something else?

aleksmaus replied:
:-D we had a big discussion on naming here, initial name for this ```agent_actions```, then the agent team asked to changed it to proxied_actions. 
Essentially it is the Agent actions that specific components what to receive. 
For Endpoint tamper protection specifically it needed signed by kibana material (in this case UNENROLL/UPGRADE actions) in order to be able to proceed with unprotecting itself before unenroll or upgrade.

aleksmaus replied:
Updated the doc. 
Leaving the naming up to @blakerouse and @cmacknz, since the current naming was already changed based on their suggestion.

blakerouse replied:
To me the actions are blocking, so they need to be properly handled on the component, before the Elastic Agent can process the action itself. I feel like `forwarded_` means that its just forwarded and not used itself. `proxied` to me means that the action is sent to the component first and then the Elastic Agent does some work itself.

More of a fan of the `proxied_actions`.

faec replied:
That's a good point actually, the documentation doesn't mention that these calls can potentially cancel the action and/or block the entire action dispatcher for a significant amount of time, can you add a bit more to make that clear?

faec reviewed (commented):
Thank you for the changes so far, a couple more questions about terminology and how the signature is used.

aleksmaus commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,167 @@
+
+	actionType := action.Type()
+
+	backExp := backoff.NewExpBackoff(ctx.Done(), d.minBackoff, d.maxBackoff)
```
this is not quite correct, will push the change today.

aleksmaus replied:
Done.

faec commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,166 @@
+	ucs := make([]unitWithComponent, 0)
+	for _, comp := range state.Components {
+		if comp.Component.InputSpec != nil && contains(comp.Component.InputSpec.Spec.ProxiedActions, typ) {
+			name := comp.Component.InputSpec.Spec.Name
```
As of [today's checkin](https://github.com/elastic/elastic-agent/pull/2952) you can get this value directly from the component with `comp.Component.InputType` instead of `comp.Component.InputSpec.Spec.Name`.

Checks of `unit.Config.Type` below are redundant since all inputs in a component must have the same type -- you could remove those checks and add a comment to this effect, or keep it as is and comment with `// All input units should match the component input type, but let's be cautious` or something ^^

aleksmaus replied:
Updated.

faec commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,166 @@
+
+// dispatchActionInParallel dispatches actions to the units/components in parallel, with exponential backoff and timeout
+func dispatchActionInParallel(ctx context.Context, log *logp.Logger, action dispatchableAction, ucs []unitWithComponent, performAction performActionFunc) error {
+	return newBackoffActionDispatcher(log, performAction).Dispatch(ctx, action, ucs)
```
What's the reasoning for making this a standalone object with a single method instead of just implementing the contents of `Dispatch` here? It looks like this object is never shared outside the package, so the public `Dispatch` is unneeded, and I worry this will be confused with the main `ActionDispatcher` which is a persistent global object that invokes all standard action handlers, where this is a one-off ephemeral struct that only exists for the duration of its call. Could we remove the struct wrapper and call this function e.g. `notifyUnitsOfProxiedAction`?

aleksmaus replied:
wanted to keep the struct to allow parameterize the backoff/timeout, especially for testing, using shorter time durations. can rename.

aleksmaus commented on internal/pkg/agent/application/actions/handlers/handler_action_unenroll.go:
```diff
@@ -60,11 +67,28 @@ func (h *Unenroll) Handle(ctx context.Context, a fleetapi.Action, acker acker.Ac
+		state := h.coord.State()
+		ucs := findMatchingUnitsByActionType(state, a.Type())
+		if len(ucs) > 0 {
+			err := dispatchActionInParallel(ctx, h.log, action, ucs, h.coord.PerformAction)
```
Added an early return of runtime PerformAction returns ```ErrNoUnit```.

faec replied:
👍 The ErrNoUnit check helps a lot with probably the most likely hiccup. I'm about to be on vacation but if @cmacknz or @blakerouse approve in my absence that's ok with me (there will be followup work but we were planning that already). If not then we can finish polishing it next week :-)

blakerouse commented on changelog/fragments/1684357852-Pass-uninstall-token-from-Agent-arguments-to-Endpoint-uninstall-command.yaml:
```diff
@@ -0,0 +1,28 @@
+kind: feature
+
+# Change summary; a 80ish characters long description of the change.
+summary: Pass uninstall token from Agent arguments to Endpoint uninstall command
```
I think we should make this clearer for users reading our release notes.

```suggestion
summary: Add new `--uninstall-token` to allow uninstall when Endpoint protection is enabled.
```

blakerouse commented on changelog/fragments/1684858962-Support-signed-UNENROLL-action-and-UNENROLL-forwarding-to-Endpoint.yaml:
```diff
@@ -0,0 +1,32 @@
+kind: feature
+
+# Change summary; a 80ish characters long description of the change.
+summary: Support signed UNENROLL action and UNENROLL forwarding to Endpoint
```
Same I think we should make this clearer.

```suggestion
summary: Enable unenroll action to complete successfully when Endpoint protection is enabled.
```

blakerouse commented on changelog/fragments/1684357852-Pass-uninstall-token-from-Agent-arguments-to-Endpoint-uninstall-command.yaml:
```diff
@@ -0,0 +1,28 @@
+# If not present is automatically filled by the tooling finding the PR where this changelog fragment has been added.
+# NOTE: the tooling supports backports, so it's able to fill the original PR number instead of the backport PR number.
+# Please provide it if you are adding a fragment for a different PR.
+pr: https://github.com/elastic/elastic-agent/pull/2708
```
Being this has been combined from other PR's, please update to this PR.

blakerouse commented on changelog/fragments/1684858962-Support-signed-UNENROLL-action-and-UNENROLL-forwarding-to-Endpoint.yaml:
```diff
@@ -0,0 +1,32 @@
+# If not present is automatically filled by the tooling finding the PR where this changelog fragment has been added.
+# NOTE: the tooling supports backports, so it's able to fill the original PR number instead of the backport PR number.
+# Please provide it if you are adding a fragment for a different PR.
+pr: https://github.com/elastic/elastic-agent/pull/2723
```
Being this has been combined from other PR's, please update to this PR.

blakerouse commented on changelog/fragments/1685542803-Tamper-protected-Endpoint-integration-removal.yaml:
```diff
@@ -0,0 +1,32 @@
+kind: feature
+
+# Change summary; a 80ish characters long description of the change.
+summary: Tamper protected Endpoint integration removal
```
Do we need the changelog entry as well? Seems rather vague, I don't know if that will help a user reading the release notes to know what this really adds.

aleksmaus replied:
I will combine these 3 change log files into one incorporating all the suggested changes.

blakerouse commented on internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go:
```diff
@@ -51,6 +51,25 @@ func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker
 		return nil
 	}
+
+	if features.TamperProtection() {
```
Why not the same pattern from the UNENROLL action with `tamperProtectionFn`?

I would prefer to see that they follow the same pattern when checking if tamper protection is enabled.

blakerouse commented on internal/pkg/agent/cmd/uninstall.go:
```diff
@@ -34,6 +34,7 @@ Unless -f is used this command will ask confirmation before performing removal.
 	}
 
 	cmd.Flags().BoolP("force", "f", false, "Force overwrite the current and do not prompt for confirmation")
+	cmd.Flags().String("uninstall-token", "", "Uninstall protection token")
```
I think some more detail on this flags description would be helpful for users running `--help`.

intxgo replied:
maybe "Uninstall token required for uninstall protected agent"

aleksmaus replied:
"Uninstall token required for protected agent uninstall"

jlind23 commented:
@blakerouse @faec could you please take a look at this again?

pierrehilbert commented:
@faec could you have a look please? @blakerouse is off for the week.

faec commented on internal/pkg/agent/application/actions/handlers/handler_action_unenroll_test.go:
```diff
@@ -0,0 +1,210 @@
+func makeComponentState(name string, proxiedActions []string) runtime.ComponentComponentState {
+	return runtime.ComponentComponentState{
+		Component: component.Component{
+			Units: []component.Unit{
```
If you add `InputType: name` to the `Component` here it should fix some of your unit test failures.

faec reviewed (commented):
Looks nearly ready (needs resync and test fix), but would still like to see description in the docs of how proxied action handlers can block / cancel the action

faec reviewed (approved):
Thanks!

cmacknz commented on internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go:
```diff
@@ -51,6 +54,25 @@ func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker
+			h.log.Debugf("handlerUpgrade: proxy/dispatch action '%+v'", a)
+			err := notifyUnitsOfProxiedAction(ctx, h.log, action, ucs, h.coord.PerformAction)
+			h.log.Debugf("handlerUpgrade: after action dispatched '%+v', err: %v", a, err)
+			if err != nil {
```
Make sure that the error here is logged at the error level so we can always debug this as a source of upgrade failures.

CC @ycombinator, we will want to make sure we have a clear way to indicate that an upgrade failed because the action was not properly signed.

cmacknz commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,174 @@
+		d.log.Debugf("Dispatch %v action to %v", actionType, uc.unit.Config.Type)
+		res, err := d.performAction(ctx, uc.component, uc.unit, uc.unit.Config.Type, params)
+		if err != nil {
+			d.log.Debugf("%v failed to dispatch to %v, err: %v", actionType, uc.component.ID, err)
```
The context here should probably wrap the error if it isn't already part of it.

cmacknz commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,174 @@
+			err := dispatch(uc)
+			if err != nil {
+				if backExp.Wait() {
+					d.log.Debugf("%v action dispatch to %v with backoff attempt: %v, after %v since start", actionType, uc.component.ID, attempt, time.Since(start))
```
Since the proxied actions types are unenroll and upgrade, and those are not frequent, it is worth bumping these logs up to the info level to make it more likely we get them by default. We can then turn them off if we find them too verbose, but I would rather start with too much logging than not enough.

cmacknz commented on pkg/component/runtime/service.go:
```diff
@@ -452,19 +571,58 @@ func (s *serviceRuntime) install(ctx context.Context) error {
 	// Always retry for internal attempts to uninstall, because they are an attempt to converge the agent's current state
 	// with its desired state based on the agent policy.
-	return uninstallService(ctx, s.log, s.comp, s.executeServiceCommandImpl, true)
+	return uninstallService(ctx, s.log, s.comp, "", s.executeServiceCommandImpl, true)
```
This will currently retry forever on failure, and after https://github.com/elastic/elastic-agent/pull/3086 it will retry for up to 1 hour.

You should mark signature validation as a permanent failure to short circuit the retry logic: https://pkg.go.dev/github.com/cenkalti/backoff/v4#Permanent

This can be done everywhere a signature failure could be retried.

cmacknz commented on internal/pkg/agent/application/actions/handlers/handler_helpers.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+const (
+	defaultActionDispatcherTimeout    = 40 * time.Second
+	defaultActionDispatcherBackoffMin = 500 * time.Millisecond
+	defaultActionDispatcherBackoffMax = 10 * time.Second
```
These could all be configurable in the agent policy, this would allow us to change them using the new Fleet agent policy override API if we find out they are too short for whatever reason.

For now, disabling the tampering feature flag is also an escape hatch for this problem.

cmacknz reviewed (approved):
Approving, thanks this was easy to follow.

A few general comments:

1. Merge this after https://github.com/elastic/elastic-agent/pull/3086 is fixed because it is going to conflict and you are better positioned to deal with conflicts properly.
2. Bias towards logging at the info level for all of the action handling since UNENROLL and UPGRADE are infrequent but important.
3. As a follow up we should extend the existing endpoint integration tests to enable tamper protection once all the pieces are ready. I don't think we need separate tests, since tamper protection is a super-set of functionality on top of the regular install and uninstall. We can just enable the feature flag with Fleet API in the tests.

aleksmaus commented:
@AndersonQ could you unblock this PR? 
@cmacknz You mention that you can unblock it if @AndersonQ is on vacation. I know you are on vacation too this week.
Maybe somebody on the agent team help? @pierrehilbert @jlind23

aleksmaus commented:
The blocking breakage on agent main branch was rolled back. I rebased this PR against the latest main branch and retested everything with tamper protection feature flag disabled/enabled and with tamper protection enabled/disabled per policy.
As mentioned before the following use cases were tested with all the combinations of the flags above:
1. Agent local uninstall command
2. Agent UNENROLL through fleet
3. Endpoint integration removal from the policy
4. Agent UPGRADE through fleet

pierrehilbert commented:
Done

amitkanfer commented:
🚀

aleksmaus commented:
> @amolnater-qasource I'm adding the `QA:Needs Validation` here as we will need you to pay special attention to any change you might see while installing and upgrading agent with Endpoint integration installed.

By default the feature is disabled (and behind feature flag) so there should be full backwards compatibility and no changes in behavior. We want to have additional validation when the feature flag is enabled before 8.11. The 8.11 is expected to have the feature enabled by default.

kevinlog commented:
> By default the feature is disabled (and behind feature flag) so there should be full backwards compatibility and no changes in behavior. We want to have additional validation when the feature flag is enabled before 8.11. The 8.11 is expected to have the feature enabled by default.

We could have @amolnater-qasource test with the flag on.  We can give instructions on how to enable both in Kibana and in the Agent Policy

aleksmaus commented:
> We could have @amolnater-qasource test with the flag on. We can give instructions on how to enable both in Kibana and in the Agent Policy

This should definitively be tested for both cases when flag is disabled and enabled.

amolnater-qasource commented:
Hi @jlind23 @kevinlog 

Thank you for the updates.

We will keep track of all the tests with Elastic Defend integration.

> We can give instructions on how to enable both in Kibana and in the Agent Policy

Could you please share the same so that we can validate this on 8.9.0 BC build.

Further could you please confirm if:
- uninstall-token will be received by running- `GET /api/fleet/uninstall_tokens` rquest in Dev tools?
- will this be possible on 8.9.0 Kibana too?

Please let us know if anything else is required from our end.

Thanks!

jlind23 commented:
> Could you please share the same so that we can validate this on 8.9.0 BC build.

This will be included only from 8.10 on and not 8.9.
I let @aleksmaus chime in for the testing instructions.

aleksmaus commented:
The testing use cases and instructions are listed in the comments to this PR.
Here is a recompiled together list of things to test.

For 8.10 release the tamper protection is behind the feature flag and is disabled by default. 
When the feature flag is disabled the Agent should behave exactly as Agent 8.9.
Before enabling the tamper protection feature flag perform the full regression test for the following use cases with and without Endpoint where appropriate.

Upgrade use case is tricky, since we need to test the upgrade from 8.9 to 8.10 (where the tamper protection was introduced initially) and also the upgrade from 8.10 to 8.11 (we had to use local builds with different versions stamped and different hashes to emulate that).

Test the following use case before enabling feature flag.

1. Uninstall the agent via command line ```sudo elastic-agent uninstall```.
2. Unenroll the agent via fleet. This should uninstall Endpoint, but leaves the unenrolled Agent running.
3. Remove Endpoint integration. This should uninstall Endpoint.
4. Upgrade the agent via fleet.


The tamper protection feature is not enabled with Kibana by default and in order to enable for Kibana add this to Kibana config:
```
xpack.fleet.enableExperimental:
  - agentTamperProtectionEnabled
```

The following comment on this PR explains how to enable the tamper protection feature flag for the Agent.
The feature is enabled per policy.
https://github.com/elastic/elastic-agent/pull/2781#issuecomment-1608446295

The purpose of the feature flag is to enable/disable the new code paths related to the tamper protection implementation. With the feature flag enabled but the tamper protection disabled test all the 4 cases above. The Endpoint should still have the tamper protection disabled pass the following tests.

1. Uninstall the agent via command line ```sudo elastic-agent uninstall```.
2. Unenroll the agent via fleet. This should uninstall Endpoint, but leaves the unenrolled Agent running.
3. Remove Endpoint integration. This should uninstall Endpoint.
4. Upgrade the agent via fleet.


Now the tamper protection can be enabled for the Endpoint via Kibana UI
<img width="775" alt="Screenshot 2023-06-12 at 2 39 26 PM" src="https://github.com/elastic/elastic-agent/assets/872351/09eedc3d-a9c0-4a6c-a347-0793fe77eebc">

See more notes here:
https://github.com/elastic/elastic-agent/pull/2781#issuecomment-1587880865

Please note that the tamper protection is not applied immeditally after you toggle that switch and save the policy. It take few moment to get the new configuration down to Endpoint and Endpoint to enable the tamper protection.

The easiest way to check that the "tamper protection" was enabled for Endpoint is to attempt to uninstall it without uninstall token, you should get the error:
```
$ sudo /opt/Elastic/Endpoint/elastic-endpoint uninstall --log stderr
2023-06-12 17:48:32: debug: Utilities.cpp:308 Protected data {"agent":{"protection":{"enabled":true,"signing_key":"MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAENxwK08uYPM8D5yTmoA/emoxK5s5lQFyt/6E4LpMmpxbukMBN3/rNfjodisQBPyVQmE/ENuTbE42G5K8iXXvBRA==","uninstall_token_hash":"tx/Vht5/TVZ7gE8LcuX0NiV9azGfpX3BkrF0/wRjFsY="}},"id":"6314e110-f8c2-11ed-ab07-d3e4064cb6ee","inputs":[{"id":"7429fcd3-075b-474e-8f03-d3dcf98fdeb8","name":"EP","revision":1,"type":"endpoint"}]}
2023-06-12 17:48:32: error: MainPosix.cpp:188 Invalid uninstall token
```

With tamper protection enabled for the policy retest these 4 cases again:

1. Uninstall the agent via command line ```sudo elastic-agent uninstall --uninstall-token 739ff6fb3f233f5d1```. The uninstall should fail if the token is missing or invalid.
2. Unenroll the agent via fleet. This should uninstall Endpoint, but leaves the unenrolled Agent running.
3. Remove Endpoint integration. This should uninstall Endpoint.
4. Upgrade the agent via fleet.    


Now, the last set of tests would be to test switching the tamper protection on and off.
So if the tamper protection was enabled and got disabled, you should be to uninstall the Agent or Endpoint without uninstall token.

Now the very last set of tests (I know I said the previous was the last set, but we have the feature flag that we need to test too) is to test the same use cases with the feature flag enabled/disabled.
From the Endpoint of view the protection is enabled when the feature flag and the tamper protection are enabled for the policy. From the Agent point of view the feature flag enables the new tamper protection code paths that are disabled completely when the feature flag is off. 

Ping me if you have any questions.

Thanks!

amolnater-qasource commented:
Hi @aleksmaus 

Thank you for the details.

We are working on this feature and have currently raised two issues on latest 8.10.0 SNAPSHOT:

- https://github.com/elastic/kibana/issues/163790
- https://github.com/elastic/elastic-agent/issues/3233

**Kibana Build details:**
```
VERSION: 8.10.0 SNAPSHOT
BUILD: 65900
COMMIT: 384a1f16c0b54acf6b501ac084024cf84b9abb5a
```

Further, we have created proposed test cases for this feature under [doc](https://docs.google.com/document/d/1csZJmDqpElF4uqlmjE3et4JT0ZpfefBp1JSa5aonOwo/edit#heading=h.kss6ib7auyd4).

Could you please review the same, and let us know if we are missing anything here.

Thanks

amolnater-qasource commented:
Hi @aleksmaus @jlind23 

We have created **30 test cases** for this feature under the Fleet test suite at the links:

- [[Windows]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is disabled.](https://elastic.testrail.io/index.php?/cases/view/271157)			
- [[Linux]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is disabled.](https://elastic.testrail.io/index.php?/cases/view/271158)			
- [[MAC]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is disabled.](https://elastic.testrail.io/index.php?/cases/view/271159)			
- [[Windows]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when feature flag is added but tamper settings are off.](https://elastic.testrail.io/index.php?/cases/view/271160)			
- [[Linux]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when feature flag is added but tamper settings are off.](https://elastic.testrail.io/index.php?/cases/view/271161)			
- [[MAC]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when feature flag is added but tamper settings are off.](https://elastic.testrail.io/index.php?/cases/view/271162)			
- [[Windows]: Validate Endpoint tests when agent tamper protection is enabled under Agent policy Settings.](https://elastic.testrail.io/index.php?/cases/view/271163)			
- [[Linux]: Validate Endpoint tests when agent tamper protection is enabled under Agent policy Settings.](https://elastic.testrail.io/index.php?/cases/view/271164)			
- [[MAC]: Validate Endpoint tests when agent tamper protection is enabled under Agent policy Settings.](https://elastic.testrail.io/index.php?/cases/view/271165)			
- [[Windows]: Validate that on disabling the Tamper protection toggle, user is able to uninstall agent without uninstall-token.](https://elastic.testrail.io/index.php?/cases/view/271166)			
- [[Linux]: Validate that on disabling the Tamper protection toggle, user is able to uninstall agent without uninstall-token.](https://elastic.testrail.io/index.php?/cases/view/271167)			
- [[MAC]: Validate that on disabling the Tamper protection toggle, user is able to uninstall agent without uninstall-token.](https://elastic.testrail.io/index.php?/cases/view/271168)			
- [[Windows]: Validate that on disabling and enabling back the Tamper protection toggle, user is not able to uninstall agent without uninstall-token.](https://elastic.testrail.io/index.php?/cases/view/271169)			
- [[Linux]: Validate that on disabling and enabling back the Tamper protection toggle, user is not able to uninstall agent without uninstall-token.](https://elastic.testrail.io/index.php?/cases/view/271170)			
- [[MAC]: Validate that on disabling and enabling back the Tamper protection toggle, user is not able to uninstall agent without uninstall-token.](https://elastic.testrail.io/index.php?/cases/view/271171)			
- [[Windows]: Validate on assigning agent to new policy with Tamper protection disabled/enabled.](https://elastic.testrail.io/index.php?/cases/view/271172)			
- [[Linux]: Validate on assigning agent to new policy with Tamper protection disabled/enabled.](https://elastic.testrail.io/index.php?/cases/view/271173)			
- [[MAC]: Validate on assigning agent to new policy with Tamper protection disabled/enabled.](https://elastic.testrail.io/index.php?/cases/view/271174)			
- [[Windows]: Validate basic agent tests on the agent with Tamper protection enabled.](https://elastic.testrail.io/index.php?/cases/view/271175)			
- [[Linux]: Validate basic agent tests on the agent with Tamper protection enabled.](https://elastic.testrail.io/index.php?/cases/view/271176)			
- [[MAC]: Validate basic agent tests on the agent with Tamper protection enabled.](https://elastic.testrail.io/index.php?/cases/view/271177)			
- [Validate Agent tamper protection toggle can be enabled or disabled under Agent policy settings.](https://elastic.testrail.io/index.php?/cases/view/271178)			
- [Validate Get uninstall command link gets enabled or disabled as per the Agent tamper protection toggle.](https://elastic.testrail.io/index.php?/cases/view/271179)			
- [Validate user get uninstall commands with uninstall-tokens for “Linux or MAC” and “Windows”.](https://elastic.testrail.io/index.php?/cases/view/271180)			
- [Validate Uninstall-tokens tab with Uninstall tokens for all agent policies is available under Fleet UI.](https://elastic.testrail.io/index.php?/cases/view/271181)			
- [Validate on clicking eye icon to show token, token for the agent policy is visible.](https://elastic.testrail.io/index.php?/cases/view/271182)			
- [Validate on hovering over actions- View uninstall command is visible and on clicking icon Uninstall commands are displayed with Uninstall token in the flyout.](https://elastic.testrail.io/index.php?/cases/view/271183)			
- [[Windows]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is removed after tests.](https://elastic.testrail.io/index.php?/cases/view/271184)			
- [[Linux]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is removed after tests.](https://elastic.testrail.io/index.php?/cases/view/271185)			
- [[MAC]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is removed after tests.](https://elastic.testrail.io/index.php?/cases/view/271186)

Please let us know if we are missing any scenarios to be covered here.

Thanks!

amolnater-qasource commented:
Hi Team,

As per our discussion with @kevinlog over slack, we would skip the execution of the feature plan associated with this feature under 8.10.0, as it will be available in 8.11.0.

Please let us know if anything else is required from our end.

Thanks!

amolnater-qasource commented:
Hi Team,

We have executed **21 testcases** under the Feature test run for the 8.11.0 release at the link:

- [Endpoint Tamper Protection](https://elastic.testrail.io/index.php?/runs/view/6169)

**Status:**

- **PASS: 21**

**Build details:**
VERSION: 8.11.0 BC4
BUILD: 68044
COMMIT: caa3613b5412552646e95d80bbbc6618d704ba09
Artifact Link: https://staging.elastic.co/8.11.0-cb971279/summary-8.11.0.html

Further, as the feature is now available without adding any feature flags, we have deprecated **09 testcases** created at https://github.com/elastic/elastic-agent/pull/2781#issuecomment-1696803114 :
- Dependent on Feature flag:
[[Windows]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is disabled.](https://elastic.testrail.io/index.php?/cases/view/271157)
[[Linux]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is disabled.](https://elastic.testrail.io/index.php?/cases/view/271158)
[[MAC]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is disabled.](https://elastic.testrail.io/index.php?/cases/view/271159)
[[Windows]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is removed after tests.](https://elastic.testrail.io/index.php?/cases/view/271184)
[[Linux]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is removed after tests.](https://elastic.testrail.io/index.php?/cases/view/271185)
[[MAC]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when Tamper protection feature flag is removed after tests.](https://elastic.testrail.io/index.php?/cases/view/271186)

- Already covered under Regression Cycle:
[[Windows]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when feature flag is added but tamper settings are off.](https://elastic.testrail.io/index.php?/cases/view/271160)
[[Linux]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when feature flag is added but tamper settings are off.](https://elastic.testrail.io/index.php?/cases/view/271161)
[[MAC]: Validate Endpoint Uninstall, Unenroll, remove endpoint integration and upgrade agent tests when feature flag is added but tamper settings are off.](https://elastic.testrail.io/index.php?/cases/view/271162)

As the testing is completed on this feature, we are marking this as **QA:Validated**.

Please let us know if anything else is required from our end.
Thanks
